### PR TITLE
Added autosave on the filters page

### LINKED ISSUE #41: Save the filters automatically without needing to press the save button
Right now, on the options page of the extension, to save the filters to the localStorage, the user have to press the save button. There is also a note present, reminding the user to press the save button for the changes to take effect. 

We can improve on this by automatically triggering the save whenever the user selects (or deselect) some checkboxes. We should wait for a second before triggering the save though, as the user might be selecting and deselecting multiple options at once and showing the notification of save again and again is not what we want.

This is to be done in the filters tab only.

The driver code is in [options.js](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/options.js) and the [helpers.js](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/helper.js) contain the logic of saving the filters to localStorage.

Also, make sure to remove the Save button and Note from the filters tab.

### PR DESCRIPTION
**Pull request by thegitone23**
Fixes #41 

**Description**
This PR aims to add auto save feature on the filters section of the options page

**Other information**

An issue currently exists and i would find it very helpful to get some suggestions on fixing the issue, chrome gave me the following error,
Capture

**Checklist:**
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Comment by makkoncept** [comment]
Hi @thegitone23, thanks for the PR. 
The error you mentioned is [described here](https://developer.chrome.com/apps/storage). The maximum number of the set, remove, or clear operations that can be performed each minute is set to 120. A `runtime.lastError` is triggered if this limit is exceeded. Because of this error, after a couple of changes in the filters section, the extension will not write those changes to the local storage. 
You should add a check for the `chrome.runtime.lastError` in the callback of `chrome.storage.sync.set` function [here](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/helper.js#L58). 
Something like the following would work
```
() => {
   if (chrome.runtime.lastError) {
        showNotification("Too many changes! Please wait some time", "negative", "snackbar-options");
   } else {
        showNotification("Settings saved!", "positive", "snackbar-options");
  }
}
```

**Review by makkoncept — CHANGES_REQUESTED** [review]
There are some linting errors that you need to be fixed before merging. 
I am tagging some of them, but you can run the command `npm run lint` to list all the errors.

File: src/options/helper.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ function addProvidersToDom(providers) {
     input.type = 'checkbox';
     input.id = providers[key];
     input.classList = 'vocab choice-field magenta-colored small-sized';
+    input.addEventListener("change",delayedSaveFiltersOptions);
```
Comment (makkoncept) [inline src/options/helper.js]: Use a single quote instead of double.
There should be a space after the comma.

File: src/options/helper.js [hunk 2]
```diff
@@ -63,6 +66,10 @@ export function saveSingleFilter(inputElements) {
   }
 }
 
+export function delayedSaveFiltersOptions() {
+  setTimeout(saveFiltersOptions, 1500);
```
Comment (makkoncept) [inline src/options/helper.js]: 'saveFilterOptions' was used before it was defined

File: src/options/helper.js [hunk 3]
```diff
@@ -45,6 +46,8 @@ export async function init() {
   restoreFilters(elements.darkModeInput);
   const providers = await getLatestProviders();
   addProvidersToDom(providers);
+  [...elements.useCaseInputs].map((x) => {x.addEventListener('change', delayedSaveFiltersOptions)});
+  [...elements.licenseInputs].map((x) => {x.addEventListener('change', delayedSaveFiltersOptions)});
```
Comment (makkoncept) [inline src/options/helper.js]: 'delayedSaveFiltersOptions' was used before it was defined.

**Comment by thegitone23** [comment]
> 
> 
> Hi @thegitone23, thanks for the PR.
> The error you mentioned is [described here](https://developer.chrome.com/apps/storage). The maximum number of the set, remove, or clear operations that can be performed each minute is set to 120. A `runtime.lastError` is triggered if this limit is exceeded. Because of this error, after a couple of changes in the filters section, the extension will not write those changes to the local storage.
> You should add a check for the `chrome.runtime.lastError` in the callback of `chrome.storage.sync.set` function [here](https://github.com/creativecommons/ccsearch-browser-extension/blob/master/src/options/helper.js#L58).
> Something like the following would work
> 
> ```
> () => {
>    if (chrome.runtime.lastError) {
>         showNotification("Too many changes! Please wait some time", "negative", "snackbar-options");
>    } else {
>         showNotification("Settings saved!", "positive", "snackbar-options");
>   }
> }
> ```

First of all thanks for the review, I fixed the linting errors and also added the message as you suggested, but as you saw the way I implemented auto save is essentially giving every checkbox save button like behavior which is a little  inefficient, can you kindly suggest some improvements over the current method.

**Comment by makkoncept** [comment]
Hi, @thegitone23. Sorry for the late response, but after testing the changes locally for some days I think adding this feature to the extension will not be user-friendly. This is because of the `MAX_WRITE_OPERATIONS_PER_MINUTE` limit. Only after a couple of changes, the limit error is shown and this will become worse in the future as we might add more data in the localStorage. The users may get confused due to this. 
We will not be adding this feature(at least for now).
Thanks for helping us out with this.

**Pull request closed without merging**